rllib/agents/ppo/ppo_tf_policy.py

```diff
@@ -193,22 +193,25 @@ def postprocess_ppo_gae(
         last_r = 0.0
     # Trajectory has been truncated -> last r=VF estimate of last obs.
     else:
-        # Input dict is provided to us automatically via the Model's
-        # requirements. It's a single-timestep (last one in trajectory)
-        # input_dict.
-        if policy.config["_use_trajectory_view_api"]:
-            # Create an input dict according to the Model's requirements.
-            input_dict = policy.model.get_input_dict(sample_batch, index=-1)
-            last_r = policy._value(**input_dict)
-        # TODO: (sven) Remove once trajectory view API is all-algo default.
+        state_in_view_req = policy.model.inference_view_requirements.get(
```

Comment:
What was the motivation for not using model.get_input_dict()? The previous code seemed much cleaner, whereas the current implementation requires the user to know a lot of details of the internal implementation. For code in policy, I would expect "trajectory view" internal implementation details to be hidden away.

Reply:
The reason for not using the model to produce the input dict (from the postprocessing sample batch) is as follows:
Let's take the simpler RNN case: The Model's view requirements contain "obs", and "state_in_0/1" (shifted from state_out_0/1 by -1). However, the input dict is needed for "next_obs" (very last one in the trajectory) and "state_out_0/1" (very last one in the trajectory). This need to create yet another action/value computation at the very end of the trajectory is specific to the Policy (in this case: PPO), not the Model!
We need to completely separate what the Model must know (it should not need to know, which algo it's being run with) and what the Policy knows (its postprocessing and loss requirements). So Model.get_input_dict cannot solve this.

The best solution is still:
- Let the Policy define: I need an (already "compiled"!) single timestep(!) input dict at the very end of the trajectory in each postprocessing batch, b/c of GAE.
- The SampleCollector will provide that input-dict within the postprocessing SampleBatch, thereby abiding to the model's view requirements (obs -> next-obs shift and state_in_0/1 -> state_out_0/1 shift) and the given trajectory index (-1).
- Due to the input dict being only for a single action/value computation at the end of the trajectory, we save having to copy next_obs and any state_outs at all. In experiments, this gave another performance boost (~20%), which we are currently missing out on.

Reply:
Hmm I see, so basically you're saying the previous implementation wasn't correct since it was using the last "obs" and not the last "next_obs"? I see how this is a problem.

However, can't we still move the code here into some method of the model so the policy code is simplified. I.e., why doesn't this work?

```
if traj_view_api:
   input_dict = policy.model.get_input_dict(sample_batch, index=MAX_T_PLUS_ONE)   # special index to get the view for the next obs after the end of the rollout
   last_r = policy._value(**input_dict)
```

Reply:
Then, the attention net calculations can still be handled in *general purpose code* and not in policy-specific code.

Reply:
Well, that's exactly what I was achieving with the previous "input_dict view" (`ViewRequirement(is_input_dict=True, index=-1)`). It would have solved the separation-of-concerns goal and makes it very efficient.

a) Make the Policy work w/o it ever knowing about the model it's using.
b) Make the model work w/o ever having to know, which algo it's being run in.
c) Leave logic dealing with ViewReqDicts inside the SampleCollector, which is the only object having direct access to still-in-progress buffers and can handle these requests more efficiently.

MAX_T_PLUS_ONE is just as hackish, isn't it? Using an int (e.g. -1) wouldn't work as it could be confused with an actual index. So a string?

Reply:
Yeah, so it seems we agree that the policy shouldn't be handling model specific code. The difference is what the general purpose code should look like.

IMO, the problem with get_input_dict was it's confusing what it means (the user has to understand what an "input dict"). I disagree with requirement (c) also, I really don't think this is perf critical and we shouldn't have to sacrifice code simplicity for this.

If you think the MAX_T_PLUS_ONE or -1 is confusing, why not add some helper method like `get_traj_end_input_dict` which is pretty easy to understand ("this returns the input dict for observation after the end of the trajectory, useful for value bootstrapping")?

Reply:
For the optimization, I think the longer-term solution is to "always" track the 1-step extra metadata in the sample batch / collectors so it's possible to calculate `get_input_dict_for_traj_end` with no extra overheads.